### ISSUE TITLE
Add multiple types of calendars in addition to AdventCalendar

### ISSUE THREAD

**Issue by Peter9192** [AI4S2S/ai4s2s#30]
Currently, we have the `AdventCalendar` as the main class in our time module. This calendar counts down to an anchor date, and divides the year into evenly sized blocks counted in days, not accounting for irregular time conventions such as seasons or months. While this is a perfectly valid approach, there might also be a use case for different calendars. Just some ideas that come to mind:

- DefaultCalendar/ForwardCalendar/DailyCalendar... : the opposite of Advent, counting forward starting from an anchor date. I think having this as a default also clarifies what's special about the AdventCalendar. 
- MonthlyCalendar: similar to DefaultCalendar/AdventCalendar, but instead of counting in days, it will count in months.
- GapCalendar: manually designate a precursor period and a target period, leaving out everything in between and outside these periods.

Perhaps we could consider renaming the `AdventCalendar` to `BackwardCalendar` to contrast it more clearly with the `ForwardCalendar`, or to `BackwardDailyCalendar` to contrast it with a `DailyCalendar`. Alternatively, we could add an argument `backwards=True` or `countdown=True` to make it possible to use the same calendar for both use cases.

To implement multiple calendar, we can make a BaseCalendar class and let the other calendars inherit from it, like so:

```Py
class BaseCalendar():
    # pass

class DailyCalendar(BaseCalendar):
    # pass

class MontlyCalendar(BaseCalendar):
    # pass

# etc.
```
The advantage is that we will be able to reuse functions that can be shared across all Calendars, and that we keep the API more or less uniform across all calendars.

**Comment by BSchilperoort** [comment]
(From _https://github.com/AI4S2S/s2spy/pull/60#issuecomment-1210674986)_

An alternative could also to have the users interface with a single calendar object, e.g. AdventCalendar (or CountdownCalendar, BackwardCalendar or any other suitable name).

Where the following would all be valid inputs;

```py
daycalendar = AdventCalendar(anchor=(31, 12), freq='4d')
weekcalendar = AdventCalendar(anchor=40, freq='2W')
monthcalendar = AdventCalendar(anchor='Jan', freq='1M')
```

And then the only two calendars for a user to remember would be the `ForwardCalendar` and `BackwardCalendar`, and if it is daily/weekly/monthly is handled internally.

**Comment by Peter9192** [comment]
So the constructor would return a different subclass depending on the input arguments? Not sure if that's a common code pattern, but I do like the clean interface

**Comment by BSchilperoort** [comment]
Yes, that's what I meant. I have been thinking of examples of this in other packages, but I also could not come up with any.

But I feel like it's either that, or having a big list of possible calendars somewhere in the docs. Or perhaps some better solution exists that we haven't thought of yet.

**Comment by Peter9192** [comment]
Perhaps the constructor then doesn't have to be a class, but it can be a simple function that returns the appropriate class. 

Here are some similar examples, for inspiration:

- [Pandas `interval_range`](https://pandas.pydata.org/docs/reference/api/pandas.interval_range.html#pandas.interval_range)
- [SKLearn `make_pipeline`](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.make_pipeline.html#sklearn.pipeline.make_pipeline)
- [SKLearn `train_test_split`](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html#sklearn.model_selection.train_test_split) slightly different, but still interesting as inspiration.
